docs/src/manuals/custom-node.md

```diff
@@ -0,0 +1,338 @@
+
+
+---
+
+
+## Step 4: Defining the average energy for our node
+
+
+To complete the computation of the Bethe free energy, we also need to compute the average energy term. The average energy in our `MyBernoulli` example can be computed as $-\mathrm{E}_{q(x_k, \pi)}[\ln p(x_k \mid \pi)]$, however, because we know that we observe $x_k$ and therefore $q(x_k, \pi)$ factorizes, we can instead compute
+$$\begin{aligned}
+-\mathrm{E}_{q(x_k)q(\pi)}[\ln p(x_k \mid \pi)]
+&= -\mathrm{E}_{q(x_k)q(\pi)} [\ln (\pi^{x_k} (1-\pi)^{1 - x_k})] \\
+&= -\mathrm{E}_{q(x_k)q(\pi)} [x_k \ln(\pi) + (1-x_k) \ln(1-\pi)] \\
+&= -\mathrm{E}_{q(x_k)}[x_k] \mathrm{E}_{q(\pi)} [\ln(\pi)] - (1-\mathrm{E}_{q(x_k)}[x_k]) \mathrm{E}_{q(\pi)}[\ln(1-\pi)]
+\end{aligned}$$
+
+Which is what we implemented below. Note that `mean(mirrorlog, q(x))` is equal to $\mathrm{E}_{q(x)}[1-\log{x}]$.
+
+```julia
+@average_energy Bernoulli (q_out::Any, q_π::Any) = -mean(q_out) * mean(log, q_π) - (1.0 - mean(q_out)) * mean(mirrorlog, q_π)
+```
+
+
+
+In the case that the interfaces do not factorize, we would get something like `@average_energy MyBernoulli (q_out_π) ...`.
+
+
+## Using our node in a model
+
+
+With all the necessary functions defined, we can proceed to test our custom node in an experiment. For this experiment, we will generate a dataset from a Bernoulli distribution with a fixed success probability of 0.75. Next, we will define a probabilistic model that has a `Beta` prior and a `MyBernoulli` likelihood. The `Beta` prior will be used to model our prior belief about the probability of success. The MyBernoulli likelihood will be used to model the generative process of the observed data. We start by generating the dataset:
+
+```julia
+using Random
+
+rng = MersenneTwister(42)
+n = 500
+π_real = 0.75
+distribution = Bernoulli(π_real)
+
+dataset = float.(rand(rng, distribution, n));
+```
+
+
+
+Next, we define our model. Note that we use the `MyBernoulli` node in the model. The model consists of a single latent variable `π`, which has a `Beta` prior and is the parameter of the `MyBernoulli` likelihood. The `MyBernoulli` node takes the value of `π` as its parameter and returns a binary observation. We set the hyperparameters of the `Beta` prior to be 4 and 8, respectively, which correspond to a distribution slightly biased towards higher values of `π`. The model is defined as follows:
+
+```julia
+@model function coin_model_mybernoulli(n)
+
+    # `datavar` creates data 'inputs' in our model
+    y = datavar(Float64, n)
+
+    # We endow θ parameter of our model with some prior
+    π ~ Beta(4.0, 8.0)
+
+    # We assume that outcome of each coin flip is governed by the MyBernoulli distribution
+    for i in 1:n
+        y[i] ~ MyBernoulli(π)
+    end
+
+end
+```
+
+
+
+Finally, we can run inference with this model and the generated dataset:
+
+```julia
+result_mybernoulli = inference(
+    model = coin_model_mybernoulli(length(dataset)), 
+    data  = (y = dataset, ),
+)
+```
+
+```
```

Comment:
You don't need to show the results yourself if you use the `@example` block (see my previous comment). It does that automatically.